docs/contracts.md

```diff
@@ -214,18 +214,160 @@ fn log_event(start: *const u8, length: u32) -> i32;
 
 ## State operations
 
-TODO: This will change for P4.
+The state exposes a key-value store with arbitrary byte arrays as values.
+Internally the tree is implemented as a radix tree with branching on 4 bits, but
+the interface that is exposed is a key-value store. Iteration is supported over
+subtrees anchored at specific prefixes.
+
+When looking up, or iterating over the tree entries are given out to the
+contract. An entry is essentially a pointer to the tree node. The reason for
+this desing is in major part because in Wasm only primitive (i64 or i32) values
+may be returned by functions. In order to make lookup possible and practical to
+wrap in Rust (and other high-level languages) we return entries, which are
+encoded as 64 bit integers. Then these entries may be used to read and write the
+value at the tree node. In concordium-std the interface that is exposed is the
+Read + Write traits (and Seek).
+
+This separation of operations into two steps also affects costs. Lookup (and
+creation and deletion) costs are based on the length of the key, but further
+operations on the entry do not depend on the length of the key, but only on the
+data that is being written or read.
+
+There is a non-trivial interaction between entries and contract calls. When a
+contract A invokes contract B, that might lead to contract A's state being
+modified (e.g., B = A, or B further calls C, ...). This means that if an entry
+is given out before the call, it might no longer be valid after the call (e.g.,
+the value at the key might have been deleted). For this reason if as a result of
+a contract call contract A's state was modified (which is defined as any of the
+state modification functions being called) then all the entries that were given
+out are invalidated and attempting to use them will lead to an error return
+value.
+
+### Lookup, creation, and deletion
+```rust
+    /// Lookup an entry with the given key. The return value is either
+    /// u64::MAX if the entry at the given key does not exist, or else
+    /// the first bit of the result is 0, and the remaining bits
+    /// are an entry identifier that may be used in subsequent calls.
+    pub(crate) fn state_lookup_entry(key_start: *const u8, key_length: u32) -> u64;
```

Comment:
I have added this, but I do not think it adds value in arguments since they are already named.

Return values are generally not entries or iterators, but some form of Option<Entry> or Result<Option<Entry>, ...>.